File: example/Android/app/src/main/kotlin/com/appreveal/example/navigation/ExampleRouter.kt
package com.appreveal.example.navigation

import com.appreveal.state.NavigationProviding

object ExampleRouter : NavigationProviding {

    override var currentRoute: String = "orders.list"
        private set

    private val routeStack: MutableList<String> = mutableListOf("orders.list")
    private val modalStackInternal: MutableList<String> = mutableListOf()

    fun push(route: String) {
        routeStack.add(route)
        currentRoute = route
    }

    fun pop() {
        if (routeStack.size > 1) {
            routeStack.removeLast()
            currentRoute = routeStack.lastOrNull() ?: "unknown"
        }
    }

    fun presentModal(route: String) {
        modalStackInternal.add(route)
    }

    fun dismissModal() {
        if (modalStackInternal.isNotEmpty()) {
            modalStackInternal.removeLast()
        }
    }

    override val navigationStack: List<String> get() = routeStack.toList()
    override val presentedModals: List<String> get() = modalStackInternal.toList()
}

File: example/macOS/AppRevealMacExample/Navigation/ExampleRouter.swift
import Foundation

#if DEBUG
import AppReveal
#endif

final class ExampleRouter {

    static let shared = ExampleRouter()

    private(set) var currentRoute = ExampleSection.orders.rootRoute
    private(set) var routeStack: [String] = [ExampleSection.orders.rootRoute]
    private(set) var modalStack: [String] = []

    private init() {}

    func switchSection(_ section: ExampleSection) {
        ExampleStateContainer.shared.selectedSection = section
        currentRoute = section.rootRoute
        routeStack = [section.rootRoute]
    }

    func showOrderDetail(_ order: ExampleOrder) {
        currentRoute = "orders.detail"
        routeStack = [ExampleSection.orders.rootRoute, "orders.detail.\(order.id)"]
    }

    func showProductDetail(_ product: ExampleProduct) {
        currentRoute = "catalog.detail"
        routeStack = [ExampleSection.catalog.rootRoute, "catalog.detail.\(product.id)"]
    }

    func presentModal(route: String) {
        modalStack.append(route)
    }

    func dismissModal(route: String) {
        modalStack.removeAll { $0 == route }
    }
}

#if DEBUG
extension ExampleRouter: NavigationProviding {
    var navigationStack: [String] { routeStack }
    var presentedModals: [String] { modalStack }
}
#endif
